### PR TITLE
Fix 14682: fix overriding check in mergeSingleDenot

### PR DESCRIPTION
**Pull request by noti0na1**
We should relax overriding check in `mergeSingleDenot` as we did in other places.

I also moved some explicit nulls tests to `tests/explicit-nulls` folder.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -767,7 +767,7 @@ object RefChecks {
           for (mbrd <- self.member(name).alternatives) {
             val mbr = mbrd.symbol
             val mbrType = mbr.info.asSeenFrom(self, mbr.owner)
-            if (!mbrType.overrides(mbrd.info, matchLoosely = true))
+            if (!mbrType.overrides(mbrd.info, false, matchLoosely = true))
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:770]:
```suggestion
            if (!mbrType.overrides(mbrd.info, relaxedCheck = false, matchLoosely = true))
```

**Review by olhotak — COMMENTED** [review]
I wonder if there is some more modular/systematic way to ensure that all the overriding checks follow the relaxed rules for Java methods. There are many places that need to check this condition and we were missing some of them; is there some way to make sure that we are not missing any now and won't miss some in the future?

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 2]
```diff
@@ -1071,21 +1071,25 @@ object Types {
...
         case _ => tp
       }
-      !checkClassInfo && this.isInstanceOf[ClassInfo]
-      || (this.widenExpr frozen_<:< that.widenExpr)
-      || matchLoosely && {
-           val this1 = widenNullary(this)
-           val that1 = widenNullary(that)
-           ((this1 `ne` this) || (that1 `ne` that)) && this1.overrides(that1, false, checkClassInfo)
-         }
+      val overrideCtx = if relaxedCheck && !ctx.mode.is(Mode.RelaxedOverriding) then ctx.relaxedOverrideContext else ctx
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala:1083]:
```suggestion
      val overrideCtx = if relaxedCheck then ctx.relaxedOverrideContext else ctx
```
`relaxedOverrideContext` already re-uses the current context if nothing changes.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala:770]: Agreed. Every true/false argument should be by name.

**Review by odersky — APPROVED** [review]
Otherwise LGTM

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 2 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: Done

**Pull request merged**